clippy_lints/src/methods/unnecessary_iter_cloned.rs

```diff
@@ -40,6 +40,68 @@ pub fn check_for_loop_iter(
         && !clone_or_copy_needed
         && let Some(receiver_snippet) = snippet_opt(cx, receiver.span)
     {
+        // Issue 12098
+        // https://github.com/rust-lang/rust-clippy/issues/12098
+        // if the assignee have `mut borrow` conflict with the iteratee
+        // the lint should not execute, former didn't consider the mut case
+
+        // check whether `expr` is mutable
+        fn is_mutable(cx: &LateContext<'_>, expr: &Expr<'_>) -> bool {
+            if let Some(hir_id) = path_to_local(expr)
+                && let Node::Pat(pat) = cx.tcx.hir_node(hir_id)
+            {
+                matches!(pat.kind, PatKind::Binding(BindingAnnotation::MUT, ..))
+            } else {
+                true
+            }
+        }
+
+        if let Some(caller) = if let ExprKind::Call(_, [child, ..]) = expr.kind {
+            // filter first layer of iterator
+            let mut child = child;
+            // get inner real caller requests for clone
+            while let ExprKind::MethodCall(_, caller, _, _) = child.kind {
+                child = caller;
+            }
+            Some(child)
+        } else {
+            None
+        } && is_mutable(cx, caller)
+        {
+            // all clones get from explicit function call
+            // the suggestions out of this scope will handle for immutable caller
+            // only check for mutable caller here
+            // if caller is mutable and caller or its fields are changed
+            // the lint should not execute
+            if if let ExprKind::Block(block, ..) = body.kind {
+                let mut no_change = true;
+                // check whether caller is mutable and caller or its fields are changed in this loop
+                for stmt in block.stmts {
+                    if let StmtKind::Semi(e) = stmt.kind {
+                        // dont check `is_mutable` for `lhs` here because assign to an immutable variable is an error
+                        if let Some(lhs) = match e.kind {
+                            ExprKind::Assign(lhs, _, _) => Some(lhs),
+                            ExprKind::AssignOp(_, lhs, _) => Some(lhs),
+                            _ => None,
+                        } {
+                            no_change = can_mut_borrow_both(cx, caller, lhs);
+                        }
+                        if !no_change {
+                            // if one change occurs, stop the loop
+                            break;
+                        }
+                    }
+                }
+                !no_change
+            } else {
+                false
+            } {
+                // skip this lint
+                return true;
+            }
+        }
```

Comment:
Instead of doing everything, we could move some part of it into a function:

```rust
fn is_caller_or_fields_changed(body: &Body) -> bool {
    if let ExprKind::Block(block, ..) = body.kind {
        let mut no_change = true;
        // check whether caller is mutable and caller or its fields are changed in this loop
        for stmt in block.stmts {
            if let StmtKind::Semi(e) = stmt.kind {
                // dont check `is_mutable` for `lhs` here because assign to an immutable variable is an error
                if let Some(lhs) = match e.kind {
                    ExprKind::Assign(lhs, _, _) => Some(lhs),
                    ExprKind::AssignOp(_, lhs, _) => Some(lhs),
                    _ => None,
                } {
                    if !can_mut_borrow_both(cx, caller, lhs) {
                        return true;
                    }
                }
            }
        }
    }
    false
}
```

Then we can just do:

```suggestion
            if is_caller_or_fields_changed(&body) {
                // skip this lint
            }
```

Reply:
Reading some more, I see that you get a temporary variable (`caller`) from your code above. We can change the call to:

```suggestion
        if let ExprKind::Call(_, [child, ..]) = expr.kind {
            // filter first layer of iterator
            let mut child = child;
            // get inner real caller requests for clone
            while let ExprKind::MethodCall(_, caller, _, _) = child.kind {
                child = caller;
            }
            if is_mutable(cx, child) && is_caller_or_fields_changed(&body, child) {
                // skip lint
                return true;
            }
        };
```

?

Reply:
Thanks! I have fix this and the CI error